File: examples/go/apps/minihermes/agent.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// AgentActor — core self-improving agent loop (message → LLM → tool_use → skills → repeat).
// Demonstrates: Ask (LLM, tools, skills, memory), KV (session history),
// TupleSpace (context tracking), PG discovery, skill injection, cron processing.
package main

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

const (
	defaultMaxIterations = 8
	defaultTokenBudget   = 8192
	compressionThreshold = 0.75 // compress when history tokens exceed 75% of budget
)

// AgentActor is the core self-improving loop. On each turn it:
//  1. Loads session history from KV
//  2. Checks context budget — compresses if >75% full
//  3. Injects matching skills into the system prompt
//  4. Runs the LLM → tool loop (max N iterations)
//  5. Persists history and evaluates whether new skills should be learned
type AgentActor struct {
	plexspaces.BaseActor
	SystemPrompt  string           `json:"system_prompt"`
	Messages      []map[string]any `json:"messages"`
	MaxHistory    int              `json:"max_history"`
	MaxIterations int              `json:"max_iterations"`
	TokenBudget   int              `json:"token_budget"`
	TotalChats    int              `json:"total_chats"`
	TotalToolCalls int             `json:"total_tool_calls"`
}

func NewAgentActor() plexspaces.Actor {
	a := &AgentActor{MaxHistory: 50, MaxIterations: defaultMaxIterations, TokenBudget: defaultTokenBudget}
	a.SetSelf(a)
	return a
}

func newAgentActor() *AgentActor {
	a := &AgentActor{MaxHistory: 50, MaxIterations: defaultMaxIterations, TokenBudget: defaultTokenBudget}
	a.SetSelf(a)
	return a
}

func (a *AgentActor) Init(configJSON string) string {
	var config struct {
		ActorID string            `json:"actor_id"`
		Args    map[string]string `json:"args"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	a.SetRuntimeMetadata(config.ActorID)
	if sp := config.Args["system_prompt"]; sp != "" {
		a.SystemPrompt = sp
	} else {
		a.SystemPrompt = "You are Hermes, a self-improving AI assistant. You learn from experience, create reusable skills, and automate recurring tasks."
	}
	if v := config.Args["max_iterations"]; v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			a.MaxIterations = n
		}
	}
	if v := config.Args["token_budget"]; v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			a.TokenBudget = n
		}
	}
	if v := config.Args["max_history"]; v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			a.MaxHistory = n
		}
	}

	if err := host.PG().Join("svc:agent"); err != nil {
		host.Warn(fmt.Sprintf("AgentActor: failed to join svc:agent: %v", err))
	}
	// Register in object registry for richer discovery (capability-aware routing)
	_ = host.Registry().Register(plexspaces.ObjectRegistration{
		ObjectID:       config.ActorID,
		ObjectType:     "actor",
		ObjectCategory: "agent",
		Capabilities:   []string{"chat", "tool_use", "skill_learning", "cron_automation", "context_compression"},
	})
	writeActorInfo(config.ActorID, "hermes-agent",
		"Self-improving agent with skill learning, tiered memory, and cron automation",
		[]string{"chat", "tool_use", "skill_learning", "cron_automation", "context_compression"})

	host.Info(fmt.Sprintf("AgentActor Init actor_id=%s max_iter=%d budget=%d", config.ActorID, a.MaxIterations, a.TokenBudget))
	return ""
}

func (a *AgentActor) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "chat":
		return a.chat(p)
	case "process_cron":
		return a.processCron(p)
	case "get_history":
		return marshal(map[string]any{"status": "ok", "messages": a.Messages, "count": len(a.Messages)})
	case "clear_history":
		sessionID := stringVal(p, "session_id", "")
		a.Messages = []map[string]any{}
		if sessionID != "" {
			host.KVDelete("session_history:" + sessionID)
		}
		return marshal(map[string]any{"status": "ok"})
	case "get_stats":
		return marshal(map[string]any{
			"status":           "ok",
			"total_chats":      a.TotalChats,
			"total_tool_calls": a.TotalToolCalls,
			"history_len":      len(a.Messages),
			"max_iterations":   a.MaxIterations,
			"token_budget":     a.TokenBudget,
		})
	default:
		return marshal(map[string]any{"error": "unknown_op", "op": op})
	}
}

func (a *AgentActor) chat(p map[string]any) string {
	userMsg := stringVal(p, "message", "")
	sessionID := stringVal(p, "session_id", "default")
	if userMsg == "" {
		return marshal(map[string]any{"error": "message is required"})
	}

	// Restore session history from KV if available
	if hist := host.KVGet("session_history:" + sessionID); hist != "" && len(a.Messages) == 0 {
		var msgs []map[string]any
		if err := json.Unmarshal([]byte(hist), &msgs); err == nil {
			a.Messages = msgs
		}
	}

	// Auto-compress if context is too large
	estimatedTokens := estimateTokens(a.Messages)
	if float64(estimatedTokens) > float64(a.TokenBudget)*compressionThreshold {
		a.compressContext(sessionID)
	}

	// Inject relevant skills into system prompt
	skillContext := a.fetchSkillContext(userMsg)

	a.Messages = append(a.Messages, map[string]any{"role": "user", "content": userMsg})

	// Get available tools
	tools := a.fetchToolSchemas()

	llmID, _ := registryFirst("llm_gateway", "svc:llm_gateway", "completion")

	loopIter := 0
	finalResponse := ""
	toolCallsThisTurn := 0

	for loopIter < a.MaxIterations {
		loopIter++

		// Build messages with system prompt + skill context
		sysContent := a.SystemPrompt
		if skillContext != "" {
			sysContent = sysContent + "\n\n## Learned Skills\n" + skillContext
		}
		msgsWithSystem := append([]map[string]any{{"role": "system", "content": sysContent}}, a.Messages...)

		if llmID == "" {
			// No LLM available: short-circuit
			finalResponse = fmt.Sprintf("Agent processed: %s (LLM not available)", userMsg)
			break
		}

		raw, err := host.Ask(llmID, "completion", map[string]any{
			"op":       "completion",
			"messages": msgsToAny(msgsWithSystem),
			"tools":    tools,
		}, 15000)
		if err != nil {
			finalResponse = fmt.Sprintf("LLM error: %v", err)
			break
		}

		llmMap, ok := raw.(map[string]any)
		if !ok {
			finalResponse = fmt.Sprintf("Agent processed: %s", userMsg)
			break
		}
		if _, hasErr := llmMap["error"]; hasErr {
			finalResponse = fmt.Sprintf("LLM unavailable: %v", llmMap["error"])
			break
		}

		resp, _ := llmMap["response"].(map[string]any)
		if resp == nil {
			finalResponse = fmt.Sprintf("Agent processed: %s", userMsg)
			break
		}

		stopReason := stringVal(resp, "stop_reason", "end_turn")
		content := stringVal(resp, "content", "")

		assistantMsg := map[string]any{"role": "assistant", "content": content, "stop_reason": stopReason}
		if tcs := resp["tool_calls"]; tcs != nil {
			assistantMsg["tool_calls"] = tcs
		}
		a.Messages = append(a.Messages, assistantMsg)

		if stopReason == "end_turn" {
			finalResponse = content
			break
		}

		if stopReason == "tool_use" {
			toolCalls, _ := resp["tool_calls"].([]any)
			toolResults := []string{}

			for _, tcRaw := range toolCalls {
				tc, ok := tcRaw.(map[string]any)
				if !ok {
					continue
				}
				tcID := stringVal(tc, "id", "tc_unknown")
				tcName := stringVal(tc, "name", "")
				tcInput, _ := tc["input"].(map[string]any)
				if tcInput == nil {
					tcInput = map[string]any{}
				}

				// Check guardrails before execution
				if a.requiresApproval(tcName) {
					a.Messages = append(a.Messages, map[string]any{
						"role":         "tool",
						"tool_call_id": tcID,
						"content":      `{"error":"tool requires approval","tool":"` + tcName + `"}`,
					})
					toolResults = append(toolResults, fmt.Sprintf("%s: requires_approval", tcName))
					continue
				}

				// Dispatch to ToolExecutorActor
				toolID, toolErr := registryFirst("tool_executor", "svc:tools")
				var toolOutput map[string]any
				if toolErr == nil {
					toolResp, askErr := host.Ask(toolID, "execute", map[string]any{
						"op":    "execute",
						"name":  tcName,
						"input": tcInput,
					}, 10000)
					if askErr == nil {
						if tr, ok := toolResp.(map[string]any); ok {
							toolOutput = tr
						}
					}
				}
				if toolOutput == nil {
					toolOutput = map[string]any{"error": "tool execution failed", "tool": tcName}
				}

				outJSON, _ := json.Marshal(toolOutput)
				a.Messages = append(a.Messages, map[string]any{
					"role":         "tool",
					"tool_call_id": tcID,
					"content":      string(outJSON),
				})
				toolResults = append(toolResults, fmt.Sprintf("%s: %s", tcName, string(outJSON)))
				toolCallsThisTurn++
				fireAudit("tool_called", fmt.Sprintf("tool=%s session=%s", tcName, sessionID))
			}

			if len(toolResults) > 0 {
				finalResponse = strings.Join(toolResults, "; ")
			}
			continue
		}

		finalResponse = content
		break
	}

	// Trim history if too long
	if len(a.Messages) > a.MaxHistory {
		keep := a.MaxHistory / 2
		a.Messages = a.Messages[len(a.Messages)-keep:]
	}

	// Persist session history
	if msgsJSON, err := json.Marshal(a.Messages); err == nil {
		host.KVPut("session_history:"+sessionID, string(msgsJSON))
	}

	// Track multi-step patterns for skill learning
	a.TotalToolCalls += toolCallsThisTurn
	if toolCallsThisTurn >= 3 {
		a.maybeLearnSkill(sessionID, toolCallsThisTurn)
	}

	a.TotalChats++
	a.IncrCounter(host, "agent_chats")
	fireAudit("agent_chat", fmt.Sprintf("session=%s iterations=%d tools=%d", sessionID, loopIter, toolCallsThisTurn))

	return marshal(map[string]any{
		"status":          "ok",
		"response":        finalResponse,
		"session_id":      sessionID,
		"loop_iterations": loopIter,
		"tool_calls":      toolCallsThisTurn,
		"messages_count":  len(a.Messages),
	})
}

// processCron dequeues and executes a cron job in an isolated session context.
func (a *AgentActor) processCron(p map[string]any) string {
	jobID := stringVal(p, "job_id", "")
	prompt := stringVal(p, "prompt", "")
	runID := fmt.Sprintf("cron-%s-%d", jobID, host.NowMs())

	if prompt == "" {
		return marshal(map[string]any{"error": "prompt is required for cron execution"})
	}

	// Execute in isolated session: clear messages first
	savedMessages := a.Messages
	a.Messages = []map[string]any{}

	cronResp := a.chat(map[string]any{
		"message":    prompt,
		"session_id": runID,
	})

	// Restore main messages; cron runs in isolation
	a.Messages = savedMessages

	// Persist cron result
	host.KVPut("cron_result:"+jobID+":"+runID, cronResp)
	fireAudit("cron_executed", fmt.Sprintf("job_id=%s run_id=%s", jobID, runID))
	host.Info(fmt.Sprintf("AgentActor: cron job executed job_id=%s run_id=%s", jobID, runID))

	var cronRespMap map[string]any
	if err := json.Unmarshal([]byte(cronResp), &cronRespMap); err == nil {
		cronRespMap["run_id"] = runID
		cronRespMap["job_id"] = jobID
		return marshal(cronRespMap)
	}
	return marshal(map[string]any{"status": "ok", "run_id": runID, "job_id": jobID})
}

// fetchSkillContext queries SkillStoreActor for relevant skills and formats them
// as a text block to inject into the system prompt.
func (a *AgentActor) fetchSkillContext(userMsg string) string {
	skillID, err := registryFirst("skill_store", "svc:skills")
	if err != nil {
		return ""
	}
	resp, askErr := host.Ask(skillID, "match_skills", map[string]any{
		"op":    "match_skills",
		"query": userMsg,
		"limit": 3,
	}, 3000)
	if askErr != nil {
		return ""
	}
	respMap, ok := resp.(map[string]any)
	if !ok {
		return ""
	}
	skills, _ := respMap["skills"].([]any)
	if len(skills) == 0 {
		return ""
	}
	lines := []string{}
	for _, s := range skills {
		sk, ok := s.(map[string]any)
		if !ok {
			continue
		}
		name := stringVal(sk, "name", "")
		procedure := stringVal(sk, "procedure", "")
		if name != "" && procedure != "" {
			lines = append(lines, fmt.Sprintf("**%s**: %s", name, procedure))
		}
	}
	return strings.Join(lines, "\n\n")
}

// fetchToolSchemas returns tool definitions from ToolExecutorActor.
func (a *AgentActor) fetchToolSchemas() []any {
	toolID, err := registryFirst("tool_executor", "svc:tools")
	if err != nil {
		return []any{}
	}
	resp, askErr := host.Ask(toolID, "list_tools", map[string]any{"op": "list_tools"}, 5000)
	if askErr != nil {
		return []any{}
	}
	respMap, ok := resp.(map[string]any)
	if !ok {
		return []any{}
	}
	tools, _ := respMap["tools"].([]any)
	return tools
}

// compressContext delegates context compression to ContextCompressorActor.
func (a *AgentActor) compressContext(sessionID string) {
	compID, err := registryFirst("context_compressor", "svc:compressor")
	if err != nil {
		// Fallback: simple truncation
		if len(a.Messages) > 10 {
			a.Messages = a.Messages[len(a.Messages)-10:]
		}
		return
	}
	msgsJSON, _ := json.Marshal(a.Messages)
	resp, askErr := host.Ask(compID, "compress", map[string]any{
		"op":         "compress",
		"messages":   string(msgsJSON),
		"session_id": sessionID,
		"keep_last":  4,
	}, 15000)
	if askErr != nil {
		host.Debug(fmt.Sprintf("AgentActor: compression failed: %v", askErr))
		return
	}
	respMap, ok := resp.(map[string]any)
	if !ok {
		return
	}
	if compressed, ok := respMap["messages"].(string); ok {
		var newMsgs []map[string]any
		if err := json.Unmarshal([]byte(compressed), &newMsgs); err == nil {
			a.Messages = newMsgs
			host.Info(fmt.Sprintf("AgentActor: context compressed session=%s new_len=%d", sessionID, len(a.Messages)))
		}
	}
}

// maybeLearnSkill signals SkillStoreActor to propose a skill from recent conversation.
func (a *AgentActor) maybeLearnSkill(sessionID string, toolCallCount int) {
	skillID, err := registryFirst("skill_store", "svc:skills")
	if err != nil {
		return
	}
	msgsJSON, _ := json.Marshal(a.Messages)
	// Fire-and-forget: skill learning is async
	_ = host.Send(skillID, "evaluate_for_learning", map[string]any{
		"op":              "evaluate_for_learning",
		"session_id":      sessionID,
		"messages":        string(msgsJSON),
		"tool_call_count": toolCallCount,
	})
}

// requiresApproval checks whether a tool requires guardrails approval.
func (a *AgentActor) requiresApproval(toolName string) bool {
	guardID, err := registryFirst("guardrails", "svc:guardrails")
	if err != nil {
		return false
	}
	resp, askErr := host.Ask(guardID, "check", map[string]any{
		"op":   "check",
		"tool": toolName,
	}, 2000)
	if askErr != nil {
		return false
	}
	respMap, ok := resp.(map[string]any)
	if !ok {
		return false
	}
	return stringVal(respMap, "decision", "allow") == "requires_approval"
}

func estimateTokens(messages []map[string]any) int {
	total := 0
	for _, m := range messages {
		if content, ok := m["content"].(string); ok {
			total += len(content) / 4
		}
	}
	return total
}

func msgsToAny(msgs []map[string]any) []any {
	result := make([]any, len(msgs))
	for i, m := range msgs {
		result[i] = m
	}
	return result
}

File: examples/python/apps/minihermes/agent.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""AgentActor — self-improving agent loop.

Conversation turn: restore KV history → compress if >75% → inject skills →
fetch tool schemas → LLM loop (max 8 iterations) → persist history →
evaluate for skill learning.
"""

import json
from plexspaces import actor, state, handler, init_handler, host
from helpers import registry_first, fire_audit, write_actor_info, ask, truncate_str

_MAX_ITER = 8
_MAX_HISTORY = 50
_TOKEN_BUDGET = 8192
_COMPRESS_THRESHOLD = 0.75


@actor
class AgentActor:
    """Self-improving agent: chat, tool calling, skill learning, cron execution."""

    system_prompt: str = state(default="You are Hermes, a self-improving AI assistant. You learn from experience, create reusable skills, and automate recurring tasks.")
    messages: list = state(default_factory=list)
    max_history: int = state(default=_MAX_HISTORY)
    max_iterations: int = state(default=_MAX_ITER)
    token_budget: int = state(default=_TOKEN_BUDGET)
    total_chats: int = state(default=0)
    total_tool_calls: int = state(default=0)
    skills_learned: int = state(default=0)
    actor_id: str = state(default="")

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        args = config.get("args", {})
        self.system_prompt = args.get("system_prompt", self.system_prompt)
        if args.get("max_iterations"):
            self.max_iterations = int(args["max_iterations"])
        if args.get("token_budget"):
            self.token_budget = int(args["token_budget"])
        host.process_groups.join("svc:agent")
        # Register in object registry for capability-aware discovery
        try:
            host.registry.register(
                ctx="",
                object_id=self.actor_id,
                object_type="actor",
                grpc_address="",
                object_category="agent",
                capabilities=["chat", "tool_use", "skill_learning", "cron_execution"],
            )
        except Exception:
            pass
        write_actor_info(self.actor_id, "hermes-agent", "Self-improving agent with skill learning and cron automation",
                         ["chat", "tool_use", "skill_learning", "cron_execution"])
        host.info(f"AgentActor init actor_id={self.actor_id}")

    @handler("chat")
    def chat(self, message: str = "", session_id: str = "") -> dict:
        if not message:
            return {"error": "message is required"}

        # Restore history from KV if session_id provided
        if session_id:
            raw = host.kv.get(f"session_history:{session_id}")
            if raw:
                try:
                    self.messages = json.loads(raw)
                except Exception:
                    pass

        self.messages.append({"role": "user", "content": message})

        # Compress if over budget
        if len(self.messages) > 10:
            self._maybe_compress(session_id)

        # Inject relevant skills
        skill_context = self._fetch_skill_context(message)

        # Get tool schemas
        tools = self._fetch_tool_schemas()

        final_response = ""
        tool_call_count = 0

        for i in range(self.max_iterations):
            llm_id, _ = registry_first("llm_gateway", fallback_group="svc:llm_gateway")
            if not llm_id:
                final_response = f"[no LLM gateway] Processed: {truncate_str(message, 60)}"
                break

            msgs_with_system = [{"role": "system", "content": self.system_prompt + skill_context}] + self.messages
            llm_resp = ask(llm_id, "completion", {"messages": msgs_with_system, "tools": tools}, 15000)
            if not llm_resp or "error" in llm_resp:
                final_response = "LLM unavailable, please try again"
                break

            response = llm_resp.get("response", {})
            stop_reason = response.get("stop_reason", "end_turn")
            content = response.get("content", "")
            tool_calls = response.get("tool_calls", [])

            assistant_msg: dict = {"role": "assistant", "content": content, "stop_reason": stop_reason}
            if tool_calls:
                assistant_msg["tool_calls"] = tool_calls
            self.messages.append(assistant_msg)

            if stop_reason == "end_turn":
                final_response = content
                break

            if stop_reason == "tool_use":
                for tc in tool_calls:
                    tc_name = tc.get("name", "")
                    tc_input = tc.get("input", {})
                    tc_id = tc.get("id", "")

                    # Guardrails check
                    if not self._check_guardrails(tc_name, tc_input):
                        self.messages.append({
                            "role": "tool",
                            "tool_call_id": tc_id,
                            "content": f"Tool {tc_name} was denied by guardrails policy.",
                        })
                        fire_audit("tool_denied", f"tool={tc_name} session={session_id}")
                        continue

                    # Execute tool
                    tool_id, _ = registry_first("tool_executor", fallback_group="svc:tools")
                    tool_output = {}
                    if not tool_id:
                        tool_output = {"error": "tool executor unavailable"}
                    else:
                        result = ask(tool_id, "execute", {"name": tc_name, "input": tc_input})
                        tool_output = result or {"error": "tool execution failed"}

                    self.messages.append({
                        "role": "tool",
                        "tool_call_id": tc_id,
                        "content": json.dumps(tool_output),
                    })
                    tool_call_count += 1
                    fire_audit("tool_called", f"tool={tc_name} session={session_id}")

                final_response = f"Completed {len(tool_calls)} tool call(s) in iteration {i + 1}"
            else:
                final_response = content
                break

        # Persist history
        if session_id and self.messages:
            try:
                host.kv.put(f"session_history:{session_id}", json.dumps(self.messages[-self.max_history:]))
            except Exception:
                pass

        # Trim in-memory history
        if len(self.messages) > self.max_history:
            self.messages = self.messages[-self.max_history:]

        self.total_chats += 1
        self.total_tool_calls += tool_call_count

        # Evaluate for skill learning
        if tool_call_count >= 3:
            self._maybe_learn_skill(session_id, tool_call_count)

        host.incr_counter("agent_chats", 1)
        fire_audit("agent_chat", f"session={session_id} tools={tool_call_count}")
        return {
            "status": "ok",
            "response": final_response,
            "session_id": session_id,
            "tool_calls": tool_call_count,
            "messages_count": len(self.messages),
        }

    @handler("process_cron")
    def process_cron(self, job_id: str = "", prompt: str = "") -> dict:
        if not job_id:
            return {"error": "job_id is required"}
        # Save and isolate conversation context
        saved_messages = list(self.messages)
        self.messages = []
        result = self.chat(message=prompt or f"Execute scheduled task: {job_id}", session_id=f"cron:{job_id}")
        self.messages = saved_messages

        import hashlib
        run_id = hashlib.md5(f"{job_id}{host.now_ms()}".encode()).hexdigest()[:8]
        fire_audit("cron_executed", f"job_id={job_id} run_id={run_id}")
        return {"status": "ok", "job_id": job_id, "run_id": run_id, "result": result.get("response", "")}

    @handler("get_history")
    def get_history(self) -> dict:
        return {"status": "ok", "messages": self.messages, "count": len(self.messages)}

    @handler("clear_history")
    def clear_history(self, session_id: str = "") -> dict:
        self.messages = []
        if session_id:
            host.kv.delete(f"session_history:{session_id}")
        return {"status": "ok"}

    @handler("get_stats")
    def get_stats(self) -> dict:
        return {
            "status": "ok",
            "total_chats": self.total_chats,
            "total_tool_calls": self.total_tool_calls,
            "skills_learned": self.skills_learned,
            "messages_count": len(self.messages),
        }

    # ------------------------------------------------------------------

    def _fetch_skill_context(self, message: str) -> str:
        skill_id, _ = registry_first("skill_store", fallback_group="svc:skills")
        if not skill_id:
            return ""
        resp = ask(skill_id, "match_skills", {"query": message, "limit": 2})
        if not resp or not resp.get("skills"):
            return ""
        parts = ["\n\n--- Relevant skills from experience ---"]
        for s in resp["skills"][:2]:
            parts.append(f"Skill: {s.get('name', '')}\n{truncate_str(s.get('procedure', ''), 200)}")
        return "\n".join(parts)

    def _fetch_tool_schemas(self) -> list:
        tool_id, _ = registry_first("tool_executor", fallback_group="svc:tools")
        if not tool_id:
            return []
        resp = ask(tool_id, "list_tools", {})
        return resp.get("tools", []) if resp else []

    def _check_guardrails(self, tool_name: str, tool_input: dict) -> bool:
        guard_id, _ = registry_first("guardrails", fallback_group="svc:guardrails")
        if not guard_id:
            return True  # Allow if guardrails unavailable
        resp = ask(guard_id, "check", {"tool": tool_name, "input": tool_input})
        if not resp:
            return True
        decision = resp.get("decision", "allow")
        return decision == "allow"

    def _maybe_compress(self, session_id: str) -> None:
        comp_id, _ = registry_first("context_compressor", fallback_group="svc:compressor")
        if not comp_id:
            return
        resp = ask(comp_id, "compress", {
            "session_id": session_id or "default",
            "messages": json.dumps(self.messages),
            "keep_last": 4,
        })
        if resp and resp.get("compressed_messages"):
            try:
                self.messages = json.loads(resp["compressed_messages"])
            except Exception:
                pass

    def _maybe_learn_skill(self, session_id: str, tool_call_count: int) -> None:
        skill_id, _ = registry_first("skill_store", fallback_group="svc:skills")
        if not skill_id:
            return
        resp = ask(skill_id, "evaluate_for_learning", {
            "session_id": session_id or "default",
            "tool_call_count": tool_call_count,
            "messages": json.dumps(self.messages[-10:]),
        })
        if resp and resp.get("action") == "learned":
            self.skills_learned += 1
